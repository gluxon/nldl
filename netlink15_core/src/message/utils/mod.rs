mod create_message_iterator;
pub use create_message_iterator::create_message_iterator;
