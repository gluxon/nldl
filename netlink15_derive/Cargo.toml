[package]
name = "netlink15_derive"
version = "0.1.0"
authors = ["Brandon Cheng <brandon.cheng@protonmail.com>"]
edition = "2018"

[lib]
proc-macro = true

[dependencies]
syn = "1.0"
proc-macro2 = "1.0.18"
quote = "1.0"
thiserror = "1.0"

[dev-dependencies]
netlink15_core = { path = "../netlink15_core" }
trybuild = "1.0"
