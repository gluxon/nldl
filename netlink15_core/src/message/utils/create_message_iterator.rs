use crate::linux::nlmsg_align;
use crate::message::raw::RawNetlinkMessage;
use crate::message::DeserializeNetlinkMessageResult;
use crate::message::NetlinkMessageResponse;
use crate::message::NetlinkMessageResponseDeserializeError;
use crate::message::NetlinkPayloadResponse;
use std::convert::TryFrom;

/// Takes a byte buffer and returns an Iterator over deserialized Netlink messages. The iterator
/// continues until there are no remaining bytes in the buffer.
///
/// This is similar to the process described in libnl here:
/// https://www.infradead.org/~tgr/libnl/doc/core.html#_parsing_a_message
///
/// This iterator intentionally does not free messages that have already been iterated over to avoid
/// re-malloc'ing remaining bytes. Memory is not released until the entire iterator is dropped.
pub fn create_message_iterator<T>(
    buf: Vec<u8>,
) -> impl Iterator<Item = DeserializeNetlinkMessageResult<T>>
where
    T: NetlinkPayloadResponse,
{
    let mut offset: usize = 0;
    let mut did_last_read_error = false;
    std::iter::from_fn(move || {
        if did_last_read_error || offset >= buf.len() {
            return None;
        }

        let read_result = next(&buf[offset..]);
        match read_result {
            Ok(success) => {
                offset += success.consumed;
                Some(Ok(success.message))
            }
            Err(err) => {
                did_last_read_error = true;
                Some(Err(err))
            }
        }
    })
}

struct NextMessageRead<T: NetlinkPayloadResponse> {
    pub message: NetlinkMessageResponse<T>,
    pub consumed: usize,
}

/// Similar to the nlmsg_next function in linux/netlink.h
fn next<T>(buf: &[u8]) -> Result<NextMessageRead<T>, NetlinkMessageResponseDeserializeError<T>>
where
    T: NetlinkPayloadResponse,
{
    let raw = RawNetlinkMessage::try_from(buf)?;
    let consumed = nlmsg_align(raw.header.len as usize);
    let message = NetlinkMessageResponse::<T>::try_from(raw)?;

    Ok(NextMessageRead { message, consumed })
}

#[cfg(test)]
mod tests {
    use super::create_message_iterator;

    #[test]
    fn stops_after_error() {
        // Start with a valid nlctrl message.
        let mut buf = vec![
            0x88, 0x00, 0x00, 0x00, 0x10, 0x00, 0x02, 0x00, 0x01, 0x00, 0x00, 0x00, 0x3A, 0xC6,
            0x00, 0x00, 0x01, 0x02, 0x00, 0x00, 0x0B, 0x00, 0x02, 0x00, 0x6E, 0x6C, 0x63, 0x74,
            0x72, 0x6C, 0x00, 0x00, 0x06, 0x00, 0x01, 0x00, 0x10, 0x00, 0x00, 0x00, 0x08, 0x00,
            0x03, 0x00, 0x02, 0x00, 0x00, 0x00, 0x08, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00,
            0x08, 0x00, 0x05, 0x00, 0x08, 0x00, 0x00, 0x00, 0x2C, 0x00, 0x06, 0x00, 0x14, 0x00,
            0x01, 0x00, 0x08, 0x00, 0x01, 0x00, 0x03, 0x00, 0x00, 0x00, 0x08, 0x00, 0x02, 0x00,
            0x0E, 0x00, 0x00, 0x00, 0x14, 0x00, 0x02, 0x00, 0x08, 0x00, 0x01, 0x00, 0x0A, 0x00,
            0x00, 0x00, 0x08, 0x00, 0x02, 0x00, 0x0C, 0x00, 0x00, 0x00, 0x1C, 0x00, 0x07, 0x00,
            0x18, 0x00, 0x01, 0x00, 0x08, 0x00, 0x02, 0x00, 0x10, 0x00, 0x00, 0x00, 0x0B, 0x00,
            0x01, 0x00, 0x6E, 0x6F, 0x74, 0x69, 0x66, 0x79, 0x00, 0x00,
        ];

        // Add the start of a new message (len = 30) but nothing else to make it invalid.
        buf.extend(&[30, 0, 0, 0]);

        let mut messages = create_message_iterator::<Vec<u8>>(buf);

        assert!(matches![messages.next(), Some(Ok(_))]);
        assert!(matches![messages.next(), Some(Err(_))]);
        assert!(matches![messages.next(), None]);
    }
}
