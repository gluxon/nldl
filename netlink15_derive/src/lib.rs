extern crate proc_macro;

use parsing::nla_type::PartitionedAttributeKinds;
use proc_macro::TokenStream;
use quote::quote;

mod parsing;

// Created following pattern from:
// https://doc.rust-lang.org/1.46.0/book/ch19-06-macros.html?highlight=procedural,macros#how-to-write-a-custom-derive-macro
#[proc_macro_derive(NetlinkAttributeSerializable, attributes(nla_type, nla_type_unknown))]
pub fn netlink_attribute_serializable_derive(input: TokenStream) -> TokenStream {
    let ast = syn::parse(input).expect("Unable to parse DeriveInput from TokenStream");
    impl_netlink_attribute_serializable(&ast).into()
}

fn impl_netlink_attribute_serializable(ast: &syn::DeriveInput) -> proc_macro2::TokenStream {
    let data_enum = match &ast.data {
        syn::Data::Enum(data_enum) => data_enum,
        _ => panic!("NetlinkAttributeSerializable derive may only be used on enums."),
    };
    let partitioned_variants = PartitionedAttributeKinds::from(data_enum)
        .expect("Failed to parse enum variants in NetlinkAttributeSerializable derive.");
    if let Some(unmarked_variant) = partitioned_variants.unmarked.first() {
        panic!(
            "Please annotate all enum variants with #[nla_type(..)]. Saw \"{}\" unannotated.",
            unmarked_variant.ident
        );
    }
    let unknown_ident = match &partitioned_variants.unknown[..] {
        [variant] => vec![variant.ident],
        [] => vec![],
        [..] => panic!(
            "Only 1 variant may be marked with #[nla_type_unknown]. Saw {}",
            partitioned_variants.unknown.len()
        ),
    };

    let name = &ast.ident;
    let (none_idents, none_nla_types) =
        partitioned_variants
            .none
            .into_iter()
            .fold((vec![], vec![]), |mut acc, attr| {
                acc.0.push(attr.ident);
                acc.1.push(attr.ty);
                acc
            });
    let (some_idents, some_nla_types) =
        partitioned_variants
            .some
            .into_iter()
            .fold((vec![], vec![]), |mut acc, attr| {
                acc.0.push(attr.ident);
                acc.1.push(attr.ty);
                acc
            });

    quote! {
        impl netlink15_core::attr::NetlinkAttributeSerializable for #name {

            fn get_type(&self) -> u16 {
                use netlink15_core::attr::NetlinkAttributeSerializable;

                match self {
                    #( Self::#none_idents => #none_nla_types, )*
                    #( Self::#some_idents(_) => #some_nla_types, )*
                    #( Self::#unknown_ident(a) => NetlinkAttributeSerializable::get_type(a), )*
                }
            }

            fn serialize_payload(&self, buf: &mut Vec<u8>) {
                use netlink15_core::message::NetlinkPayloadRequest;
                use netlink15_core::attr::NetlinkAttributeSerializable;

                match self {
                    #( Self::#none_idents => {}, )*
                    #( Self::#some_idents(val) => NetlinkPayloadRequest::serialize(val, buf), )*
                    #( Self::#unknown_ident(a) => NetlinkAttributeSerializable::serialize_payload(a, buf), )*
                }
            }
        }
    }
}
