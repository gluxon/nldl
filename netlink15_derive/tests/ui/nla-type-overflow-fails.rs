use netlink15_derive::NetlinkAttributeDeserializable;
use netlink15_derive::NetlinkAttributeSerializable;
use netlink15_core::utils::ParseNlaIntError;

#[derive(Debug, PartialEq, NetlinkAttributeSerializable, NetlinkAttributeDeserializable)]
#[netlink15(deserialize(error = "ParseNlaIntError"))]
enum ControllerAttributeOperation {
    #[nla_type(0)]
    Unspec,
    #[nla_type(u16::MAX + 1)]
    Id(u32),
}

fn main() {}
