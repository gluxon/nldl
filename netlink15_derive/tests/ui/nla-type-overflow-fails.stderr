error[E0080]: evaluation of constant value failed
  --> tests/ui/nla-type-overflow-fails.rs:10:16
   |
10 |     #[nla_type(u16::MAX + 1)]
   |                ^^^^^^^^^^^^ attempt to compute `u16::MAX + 1_u16`, which would overflow
